### ISSUE TITLE
I have some problems with my service

### ISSUE THREAD

**Issue by eliassama** [qdrant/qdrant#2774]
## Current Behavior
The environment is as follows:
1. I have three machines deployed in a distributed cluster. A single machine is 16C 32G
2. Version: v1.1.1
3. There are 33 collection
4. Each set is a 512-dimensional vector, with about 10 million of them
5. collection info is :
```json
{
    "result":{
        "status":"green",
        "optimizer_status":"ok",
        "vectors_count":3977033,
        "indexed_vectors_count":3968353,
        "points_count":3977033,
        "segments_count":4,
        "config":{
            "params":{
                "vectors":{
                    "size":512,
                    "distance":"Cosine"
                },
                "shard_number":1,
                "replication_factor":2,
                "write_consistency_factor":1,
                "on_disk_payload":true
            },
            "hnsw_config":{
                "m":16,
                "ef_construct":100,
                "full_scan_threshold":10000,
                "max_indexing_threads":0,
                "on_disk":true,
                "payload_m":16
            },
            "optimizer_config":{
                "deleted_threshold":0.2,
                "vacuum_min_vector_number":1000,
                "default_segment_number":0,
                "max_segment_size":null,
                "memmap_threshold":20000,
                "indexing_threshold":20000,
                "flush_interval_sec":5,
                "max_optimization_threads":1
            },
            "wal_config":{
                "wal_capacity_mb":32,
                "wal_segments_ahead":0
            },
            "quantization_config":null
        },
        "payload_schema":{
            "user_id":{
                "data_type":"keyword",
                "points":3977033
            }
        }
    },
    "status":"ok",
    "time":0.004815987
}
```
6. config is:
```yaml
storage:
  # Where to store all the data
  storage_path: /disk/storage

  # Where to store snapshots
  snapshots_path: /disk/snapshots

  # If true - point's payload will not be stored in memory.
  # It will be read from the disk every time it is requested.
  # This setting saves RAM by (slightly) increasing the response time.
  # Note: those payload values that are involved in filtering and are indexed - remain in RAM.
  on_disk_payload: true

  # Write-ahead-log related configuration
  wal:
    # Size of a single WAL segment
    wal_capacity_mb: 32

    # Number of WAL segments to create ahead of actual data requirement
    wal_segments_ahead: 0

  # Normal node - receives all updates and answers all queries
  node_type: "Normal"

  # Listener node - receives all updates, but does not answer search/read queries
  # Useful for setting up a dedicated backup node
  # node_type: "Listener"

  performance:
    # Number of parallel threads used for search operations. If 0 - auto selection.
    max_search_threads: 0
    # Max total number of threads, which can be used for running optimization processes across all collections.
    # Note: Each optimization thread will also use `max_indexing_threads` for index building.
    # So total number of threads used for optimization will be `max_optimization_threads * max_indexing_threads`
    max_optimization_threads: 1

  optimizers:
    # The minimal fraction of deleted vectors in a segment, required to perform segment optimization
    deleted_threshold: 0.2

    # The minimal number of vectors in a segment, required to perform segment optimization
    vacuum_min_vector_number: 1000

    # Target amount of segments optimizer will try to keep.
    # Real amount of segments may vary depending on multiple parameters:
    #  - Amount of stored points
    #  - Current write RPS
    #
    # It is recommended to select default number of segments as a factor of the number of search threads,
    # so that each segment would be handled evenly by one of the threads.
    # If `default_segment_number = 0`, will be automatically selected by the number of available CPUs
    default_segment_number: 0

    # Do not create segments larger this size (in KiloBytes).
    # Large segments might require disproportionately long indexation times,
    # therefore it makes sense to limit the size of segments.
    #
    # If indexation speed have more priority for your - make this parameter lower.
    # If search speed is more important - make this parameter higher.
    # Note: 1Kb = 1 vector of size 256
    # If not set, will be automatically selected considering the number of available CPUs.
    max_segment_size_kb: null

    # Maximum size (in KiloBytes) of vectors to store in-memory per segment.
    # Segments larger than this threshold will be stored as read-only memmaped file.
    # To enable memmap storage, lower the threshold
    # Note: 1Kb = 1 vector of size 256
    # If not set, mmap will not be used.
    memmap_threshold_kb: null

    # Maximum size (in KiloBytes) of vectors allowed for plain index.
    # Default value based on https://github.com/google-research/google-research/blob/master/scann/docs/algorithms.md
    # Note: 1Kb = 1 vector of size 256
    indexing_threshold_kb: 20000

    # Interval between forced flushes.
    flush_interval_sec: 5
    
    # Max number of threads, which can be used for optimization per collection.
    # Note: Each optimization thread will also use `max_indexing_threads` for index building.
    # So total number of threads used for optimization will be `max_optimization_threads * max_indexing_threads`
    # If `max_optimization_threads = 0`, optimization will be disabled.
    max_optimization_threads: 1

  # Default parameters of HNSW Index. Could be overridden for each collection individually
  hnsw_index:
    # Number of edges per node in the index graph. Larger the value - more accurate the search, more space required.
    m: 16
    # Number of neighbours to consider during the index building. Larger the value - more accurate the search, more time required to build index.
    ef_construct: 100
    # Minimal size (in KiloBytes) of vectors for additional payload-based indexing.
    # If payload chunk is smaller than `full_scan_threshold_kb` additional indexing won't be used -
    # in this case full-scan search should be preferred by query planner and additional indexing is not required.
    # Note: 1Kb = 1 vector of size 256
    full_scan_threshold_kb: 10000
    # Number of parallel threads used for background index building. If 0 - auto selection.
    max_indexing_threads: 0
    # Store HNSW index on disk. If set to false, index will be stored in RAM. Default: false
    on_disk: true
    # Custom M param for hnsw graph built for payload index. If not set, default M will be used.
    payload_m: null

service:

  # Maximum size of POST data in a single request in megabytes
  max_request_size_mb: 64

  # Number of parallel workers used for serving the api. If 0 - equal to the number of available cores.
  # If missing - Same as storage.max_search_threads
  max_workers: 0

  # Host to bind the service on
  host: 0.0.0.0

  # HTTP port to bind the service on
  http_port: 6333

  # gRPC port to bind the service on.
  # If `null` - gRPC is disabled. Default: null
  grpc_port: 6334
  # Uncomment to enable gRPC:
  # grpc_port: 6334

  # Enable CORS headers in REST API.
  # If enabled, browsers would be allowed to query REST endpoints regardless of query origin.
  # More info: https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS
  # Default: true
  enable_cors: true

cluster:
  # Use `enabled: true` to run Qdrant in distributed deployment mode
  enabled: true

  # Configuration of the inter-cluster communication
  p2p:
    # Port for internal communication between peers
    port: 6335

  # Configuration related to distributed consensus algorithm
  consensus:
    # How frequently peers should ping each other.
    # Setting this parameter to lower value will allow consensus
    # to detect disconnected nodes earlier, but too frequent
    # tick period may create significant network and CPU overhead.
    # We encourage you NOT to change this parameter unless you know what you are doing.
    tick_period_ms: 100


# Set to true to prevent service from sending usage statistics to the developers.
# Read more: https://qdrant.tech/documentation/telemetry
telemetry_disabled: true
```

The questions are as follows:
1. Of the three machines, one machine is always OOM, and one machine uses almost no memory. There's one other machine that works.
2. The following errors are frequently displayed in logs
![image](https://github.com/qdrant/qdrant/assets/79587688/6879c7ec-91dd-4735-8f5b-e8cb0e877d81)

## Expected Behavior

1. No exception log is generated
2. Memory load balancing among the three nodes does not stop because of the OOM node, or the memory usage is small

**Comment by timvisee** [comment]
Thank you for the report.

First of all, you mention using v1.1.1. Please update to our latest version first, because there have been a lot of fixes. You should update in rolling manner by each minor version, so: v1.2.2 -> v1.3.2 -> v1.4.1 -> v1.5.1

Am I correct that this is a self hosted instance? In that case, collections (shards) are not automatically balanced over  the cluster. You can instruct nodes to transfer shards for you though to arrange this yourself. See: https://qdrant.tech/documentation/guides/distributed_deployment/#cluster-scaling

If you'd like to reduce memory usage for collections, you may consider to enable memory mapping by setting the appropriate `on_disk` flags described [here](https://qdrant.tech/documentation/tutorials/optimize/) and [here](https://qdrant.tech/documentation/tutorials/bulk-upload/#upload-directly-to-disk).

**Comment by eliassama** [comment]
> Thank you for the report.
> 
> First of all, you mention using v1.1.1. Please update to our latest version first, because there have been a lot of fixes. You should update in rolling manner by each minor version, so: v1.2.2 -> v1.3.2 -> v1.4.1 -> v1.5.1
> 
> Am I correct that this is a self hosted instance? In that case, collections (shards) are not automatically balanced over the cluster. You can instruct nodes to transfer shards for you though to arrange this yourself. See: https://qdrant.tech/documentation/guides/distributed_deployment/#cluster-scaling
> 
> If you'd like to reduce memory usage for collections, you may consider to enable memory mapping by setting the appropriate `on_disk` flags described [here](https://qdrant.tech/documentation/tutorials/optimize/) and [here](https://qdrant.tech/documentation/tutorials/bulk-upload/#upload-directly-to-disk).

thanks, i'll try it.

**Comment by agourlay** [comment]
@eliassama Any updates to share following the version upgrade?

**Comment by eliassama** [comment]
> @eliassama Any updates to share following the version upgrade?

Sorry for replying so late.

We did not upgrade the original cluster, but built a new cluster and migrated the data there.

During the migration process, we set the original cluster to read-only status and no longer write to the original cluster.

During the migration process, we perform searches on both the new cluster and the original cluster.

After the cluster migration is completed, we will take the original cluster offline.

Overall, there seems to be no experience to draw from.

We use scripts to read the data from the original cluster and insert it into the new cluster.

This way we avoid having to deal with problems that may arise during the upgrade.

However, I noticed that it seems that the file configuration of 1.6.0 has some changes compared to v1.1.1.

And as of 1.6.0, there is a webUI that can be used to perform operations directly.
